src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs

```diff
@@ -190,15 +200,73 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
 
             return null;
         }
-        
+
+        private void EnsureAtomicDocumentWrites(ClusterOperationContext context, long index)
+        {
+            if (SerializedDatabaseCommands.TryGet(nameof(DatabaseCommands), out BlittableJsonReaderArray commands))
+            {
+                ClusterCommands ??= new List<ClusterTransactionDataCommand>();
+                foreach (BlittableJsonReaderObject dbCmd in commands)
+                {
+                    var cmdType = dbCmd[nameof(ClusterTransactionDataCommand.Type)].ToString();
+                    var docId = dbCmd[nameof(ClusterTransactionDataCommand.Id)].ToString();
+                    var atomicGuardKey = "rvn-atomic-guard-" + docId;
+                    var document = (BlittableJsonReaderObject)dbCmd[nameof(ClusterTransactionDataCommand.Document)];
+                    switch (cmdType)
+                    {
+                        case nameof(CommandType.PUT):
+                            if (document.TryGetWithoutThrowingOnError(Constants.Documents.Metadata.Key,
+                                    out BlittableJsonReaderObject metadata) == false ||
+                                metadata.TryGetWithoutThrowingOnError(Constants.Documents.Metadata.ClusterTransactionIndex,
+                                    out long putIndex) == false)
+                            {
+                                putIndex = 0; // new cmp xchg entry, so no need to take an action 
+                            }
+
+                            ClusterCommands.Add(new ClusterTransactionDataCommand
+                            {
+                                Type = CommandType.CompareExchangePUT,
+                                Id = atomicGuardKey,
+                                Index = putIndex,
+                                Document = context.ReadObject(new DynamicJsonValue {["Id"] = docId}, "cmp-xchg-content")
+                            });
+                            if (metadata != null) // just to be on the safe side, should never happen
+                            {
+                                metadata.Modifications = new DynamicJsonValue(metadata) {[Constants.Documents.Metadata.ClusterTransactionIndex] = index};
+                            }
+
+                            break;
+                        case nameof(CommandType.DELETE):
+                            if (document.TryGetWithoutThrowingOnError(Constants.Documents.Metadata.ClusterTransactionIndex,
+                                out long deleteIndex))
+                            {
+                                // this can happen if the user tried to delete without first loading the document
+                                // we can safely assume that such a blind write doesn't care about the state of the document
+                                // and just proceed normally
+                                deleteIndex = -1;
```

Comment:
This is actually a problem.. I think.

If I delete and modify concurrently I might get successful response for both, while the document might or might not exist and we back to the original issue.

I think it would be better to enforce loading the document first (check on the client side as well?).

Reply:
If you call `Delete(string id)`, that is a good indication that you don't care what the current value is.